consensus/src/quorum_store/types.rs

```diff
@@ -72,143 +32,157 @@ impl PersistedValue {
+            payload: BatchPayload::new(
+                value
+                    .maybe_payload
+                    .ok_or_else(|| anyhow::anyhow!("Payload not exist"))?,
+            ),
+        })
+    }
 }
 
-impl FragmentInfo {
-    fn new(
-        epoch: u64,
-        batch_id: BatchId,
-        fragment_id: usize,
-        fragment_payload: Vec<SerializedTransaction>,
-        maybe_expiration: Option<LogicalTime>,
-    ) -> Self {
-        Self {
-            epoch,
-            batch_id,
-            fragment_id,
-            payload: fragment_payload,
-            maybe_expiration,
-        }
+#[derive(Clone, Debug, Deserialize, Serialize, CryptoHasher)]
+pub struct BatchPayload {
+    txns: Vec<SignedTransaction>,
+    #[serde(skip)]
+    num_bytes: OnceCell<usize>,
+}
+
+impl CryptoHash for BatchPayload {
+    type Hasher = BatchPayloadHasher;
+
+    fn hash(&self) -> HashValue {
+        let mut state = Self::Hasher::new();
+        let bytes = bcs::to_bytes(&self).expect("Unable to serialize batch payload");
+        self.num_bytes.get_or_init(|| bytes.len());
+        state.update(&bytes);
+        state.finish()
     }
+}
 
-    pub fn into_transactions(self) -> Vec<SerializedTransaction> {
-        self.payload
+impl BatchPayload {
+    pub fn new(txns: Vec<SignedTransaction>) -> Self {
+        Self {
+            txns,
+            num_bytes: OnceCell::new(),
+        }
     }
 
-    pub fn fragment_id(&self) -> usize {
-        self.fragment_id
+    pub fn into_transactions(self) -> Vec<SignedTransaction> {
+        self.txns
     }
 
-    pub fn batch_id(&self) -> BatchId {
-        self.batch_id
+    pub fn num_txns(&self) -> usize {
+        self.txns.len()
     }
 
-    pub fn maybe_expiration(&self) -> Option<LogicalTime> {
-        self.maybe_expiration
+    pub fn num_bytes(&self) -> usize {
+        *self
+            .num_bytes
+            .get_or_init(|| bcs::serialized_size(&self).expect("unable to serialize batch payload"))
     }
 }
-
 #[derive(Clone, Debug, Deserialize, Serialize)]
-pub struct Fragment {
-    source: PeerId,
-    fragment_info: FragmentInfo,
+pub struct Batch {
+    batch_info: BatchInfo,
+    payload: BatchPayload,
 }
 
-impl Fragment {
+impl Batch {
     pub fn new(
-        epoch: u64,
         batch_id: BatchId,
-        fragment_id: usize,
-        fragment_payload: Vec<SerializedTransaction>,
-        maybe_expiration: Option<LogicalTime>,
-        peer_id: PeerId,
+        payload: Vec<SignedTransaction>,
+        expiration: LogicalTime,
+        batch_author: PeerId,
     ) -> Self {
-        let fragment_info = FragmentInfo::new(
-            epoch,
+        let payload = BatchPayload::new(payload);
+        let batch_info = BatchInfo::new(
+            batch_author,
             batch_id,
-            fragment_id,
-            fragment_payload,
-            maybe_expiration,
+            expiration,
+            payload.hash(),
+            payload.num_txns() as u64,
```

Comment:
This is very very important that we set it based on actual payload and not based on the batchinfo that we may have received from the client. It can be an easy mistake to make - I wonder if we could either add a test easily, or compare batch_info (at least these fields) that we compute to what we received, and ignore if they don't match?

Or assuming Batch is the thing we send to get SignedDigest now, it doesn't make sense to have num_bytes and num_txns in Batch? they will be computed, set here, and send back as a part of SignedDigest? So maybe we need two different BatchInfo's or optional fields (that we can check are None otherwise ignore).

Reply:
Yes, I think this one is actually relevant to this specific PR - with fragments I don't think we were sending num_bytes and num_txns, but now we are sending BatchInfo with Batch and BatchInfo contains these numbers - but it shouldn't - because they anyway need to be recomputed by the receiver for quota and other purposes (blindly trusting would allow malicious validator to use more than quota space for instance) - currently it is done (above) and the provided values are ignored- but sending them opens up a possibility for a future bug.

Reply:
Resolving this because Payload re-computes numbers in Batch::new constructor on the receiver side, and verify ensures that it was provided correctly. I suppose there was no need to send them to start with - but it allows re-using BatchInfo structure instead of different structs (and just adding a few lines in verify to protect against) - seems fair enough for code complexity trade-off, I probably wouldn't send, but I don't have a strong opinion